private/source.rkt

```diff
@@ -82,20 +93,33 @@
     (define stx (source-read-syntax code))
     (define current-expand-observe (dynamic-require ''#%expobs 'current-expand-observe))
     (define visits-by-location (make-hash))
-    (define/guard (observe-event! sig val)
-      (guard (and (equal? sig 'visit) (syntax? val) (syntax-original? val)) else
-        (void))
-      (define loc (syntax-source-location val))
-      (guard (hash-has-key? visits-by-location loc) then
-        (void))
-      (hash-set! visits-by-location loc val))
+    (define others-by-location (make-hash))
+    (define/guard (add-original-location! hsh stx)
+      (guard (and (syntax? stx) (syntax-original? stx)) else (void))
+      (define loc (syntax-source-location stx))
+      (guard (hash-has-key? hsh loc) then (void))
+      (hash-set! hsh loc stx))
+    (define/match (observe-event! sig val)
+      [('visit val)
+       (add-original-location! visits-by-location val)]
+      [('letX-renames (list-rest trans-idss _ val-idss _))
+       (for* ([trans-ids (in-list trans-idss)] [trans-id (in-list trans-ids)])
+         (add-original-location! others-by-location trans-id))
+       (for* ([val-ids (in-list val-idss)] [val-id (in-list val-ids)])
+         (add-original-location! others-by-location val-id))]
+      [(_ _) (void)])
     (parameterize ([current-expand-observe observe-event!])
       (expand stx))
+    (define (scopes-by-location stx)
+      (define loc (syntax-source-location stx))
+      (hash-ref visits-by-location loc (λ () (hash-ref others-by-location loc #f))))
```

Comment:
Instead of storing a function in the source code analysis results, I think it might be clearer to store (an immutable copy of) the hash table itself. That way, it's a bit easier to understand what data is being captured and represented by this part of the analysis results.

Reply:
What's wrong with a function? The places that use `scopes-by-location` in other files would not benefit from having a hash table instead. When it's encapsulated in a function, the implementation details of how the data is stored are not their concern.

If you're worried about clarity and understanding, documenting the signature and purpose of the function should accomplish that better than turning it into a hash table, in my opinion.

Reply:
Functions can do anything. They can have side effects, be nondeterministic, be unexpectedly expensive to compute, etc. If the purpose of this component of the analysis is to map source locations to identifiers with all their scopes added, just give people that map. Documenting the signature and purpose doesn't sway me either since you can also document the signature and purpose of the hash map.

Reply:
I have turned it into a hash table